web/packages/shared/components/UnifiedResources/shared/guessAppIcon.ts

```diff
@@ -0,0 +1,117 @@
+/**
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import {
+  ResourceIconName,
+  resourceIconSpecs,
+  iconNames,
+} from 'design/ResourceIcon/resourceIconSpecs';
+
+import { UnifiedResourceApp } from '../types';
+
+export function guessAppIcon(resource: UnifiedResourceApp): ResourceIconName {
+  const { awsConsole = false, name, friendlyName, labels } = resource;
+
+  if (awsConsole) {
+    return 'aws';
+  }
+
+  // Label matching takes precedence and we can assume it can be a direct lookup
+  // since we expect a certain format.
+  const labelIconValue = labels?.find(l => l.name === 'teleport.icon')?.value;
+  if (labelIconValue && resourceIconSpecs[labelIconValue]) {
+    return labelIconValue as ResourceIconName;
+  }
+
+  const app = {
+    name: withoutWhiteSpaces(name)?.toLocaleLowerCase(),
+    friendlyName: withoutWhiteSpaces(friendlyName)?.toLocaleLowerCase(),
+  };
+
+  // Try a direct lookup first.
+  if (resourceIconSpecs[app.name]) {
+    return app.name as ResourceIconName;
+  }
+  if (app.friendlyName && resourceIconSpecs[app.friendlyName]) {
+    return app.friendlyName as ResourceIconName;
+  }
+
+  // Try matching by iterating through all the icon names
+  for (let i = 0; i < iconNames.length; i++) {
```

Comment:
I'm not sure, why do we need to iterate over all icons?
Can't we just do:
```ts
    if (match('adobe', app)) {
      if (match('creative', app)) return 'adobecreativecloud';
      if (match('marketo', app)) return 'adobemarketo';
      return 'adobe'; // generic
    }
```

Reply:
hrm, i wasn't really sure what you meant. 

do you mean that we should pull out the if statements out of the iteration block?

or did you mean why isn't direct look up sufficient? if this, it's because there could be other words in the name or friendly name

this is from our platform cluster (ones with spaces are friendly name, ones with dashes is name)

<img width="421" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/12c366f6-dd8b-4d55-a331-512da477772e">

Reply:
Oh, you answered 2 weeks ago, I completely missed that.

I meant pulling out the `if` statements from the loop. If know that we need an icon for `adobe`, what going through all icons gives us? That only protect us from not having `adobe` in the icon set, but I don't think we need it.
Here is a patch:
```
Index: web/packages/shared/components/UnifiedResources/shared/guessAppIcon.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/web/packages/shared/components/UnifiedResources/shared/guessAppIcon.ts b/web/packages/shared/components/UnifiedResources/shared/guessAppIcon.ts
--- a/web/packages/shared/components/UnifiedResources/shared/guessAppIcon.ts	(revision 8b9af8956fc7a3361ee33ce61605185e7a93dc65)
+++ b/web/packages/shared/components/UnifiedResources/shared/guessAppIcon.ts	(date 1721642325511)
@@ -51,48 +51,40 @@
     return app.friendlyName as ResourceIconName;
   }
 
-  // Try matching by iterating through all the icon names
-  for (let i = 0; i < iconNames.length; i++) {
-    const iconName = iconNames[i];
-
-    // Help match brands with sub brands:
-    if (iconName === 'adobe' && match('adobe', app)) {
-      if (match('creative', app)) return 'adobecreativecloud';
-      if (match('marketo', app)) return 'adobemarketo';
-      return 'adobe'; // generic
-    }
-    if (iconName === 'atlassian' && match('atlassian', app)) {
-      if (match('bitbucket', app)) return 'atlassianbitbucket';
-      if (match('jiraservice', app)) return 'atlassianjiraservice';
-      if (match('status', app)) return 'atlassianstatus';
-      return 'atlassian'; // generic
-    }
-    if (iconName === 'google' && match('google', app)) {
-      if (match('analytic', app)) return 'googleanalytics';
-      if (match('calendar', app)) return 'googlecalendar';
-      if (match('cloud', app)) return 'googlecloud';
-      if (match('drive', app)) return 'googledrive';
-      if (match('tag', app)) return 'googletag';
-      if (match('voice', app)) return 'googlevoice';
-      return 'google'; // generic
-    }
-    if (iconName === 'microsoft' && match('microsoft', app)) {
-      if (match('excel', app)) return 'microsoftexcel';
-      if (match('drive', app)) return 'microsoftonedrive';
-      if (match('note', app)) return 'microsoftonenote';
-      if (match('outlook', app)) return 'microsoftoutlook';
-      if (match('powerpoint', app)) return 'microsoftpowerpoint';
-      if (match('team', app)) return 'microsoftteams';
-      if (match('word', app)) return 'microsoftword';
-      return 'microsoft'; // generic
-    }
+  // Help match brands with sub brands:
+  if (match('adobe', app)) {
+    if (match('creative', app)) return 'adobecreativecloud';
+    if (match('marketo', app)) return 'adobemarketo';
+    return 'adobe'; // generic
+  }
+  if (match('atlassian', app)) {
+    if (match('bitbucket', app)) return 'atlassianbitbucket';
+    if (match('jiraservice', app)) return 'atlassianjiraservice';
+    if (match('status', app)) return 'atlassianstatus';
+    return 'atlassian'; // generic
+  }
+  if (match('google', app)) {
+    if (match('analytic', app)) return 'googleanalytics';
+    if (match('calendar', app)) return 'googlecalendar';
+    if (match('cloud', app)) return 'googlecloud';
+    if (match('drive', app)) return 'googledrive';
+    if (match('tag', app)) return 'googletag';
+    if (match('voice', app)) return 'googlevoice';
+    return 'google'; // generic
+  }
+  if (match('microsoft', app)) {
+    if (match('excel', app)) return 'microsoftexcel';
+    if (match('drive', app)) return 'microsoftonedrive';
+    if (match('note', app)) return 'microsoftonenote';
+    if (match('outlook', app)) return 'microsoftoutlook';
+    if (match('powerpoint', app)) return 'microsoftpowerpoint';
+    if (match('team', app)) return 'microsoftteams';
+    if (match('word', app)) return 'microsoftword';
+    return 'microsoft'; // generic
+  }
 
-    if (match(iconName, app)) {
-      return iconName;
-    }
-  }
-
-  return 'application'; // couldn't match anything
+  const matchingIcon = iconNames.find(iconName => match(iconName, app));
+  return matchingIcon || 'application'; // couldn't match anything
 }
 
 function match(
```

Reply:
thanks, i originally had it outside... but i don't remember why I put it in the loop anymore 🤷‍♀️